github/git_refs.go

```diff
@@ -58,13 +61,60 @@ func (s *GitService) GetRef(ctx context.Context, owner string, repo string, ref
 
 	r := new(Reference)
 	resp, err := s.client.Do(ctx, req, r)
-	if err != nil {
+
+	if _, ok := err.(*json.UnmarshalTypeError); ok {
+		// Multiple refs, means there wasn't an exact match.
+		return nil, resp, nil
+	} else if err != nil {
 		return nil, resp, err
 	}
 
 	return r, resp, nil
 }
 
+// GetRefs fetches a slice of Reference objects for a given Git ref.
+// If a Git ref does not match exactly, GitHub will return all refs that match it as a prefix.
+// Example:
+// feature -> [featureA, featureB]
+// featureA -> [featureA]
+//
+// GitHub API docs: https://developer.github.com/v3/git/refs/#get-a-reference
+func (s *GitService) GetRefs(ctx context.Context, owner string, repo string, ref string) ([]*Reference, *Response, error) {
+	ref = strings.TrimPrefix(ref, "refs/")
+	u := fmt.Sprintf("repos/%v/%v/git/refs/%v", owner, repo, ref)
+	req, err := s.client.NewRequest("GET", u, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	buf := new(bytes.Buffer)
+	resp, err := s.client.Do(ctx, req, buf)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	rbuf := bytes.NewReader(buf.Bytes())
+
+	var rs []*Reference
+	r := new(Reference)
+	// prioritize the most common case: a single returned ref
+	err = json.NewDecoder(rbuf).Decode(r)
+	if err == nil {
+		rs = append(rs, r)
+	} else if _, ok := err.(*json.UnmarshalTypeError); ok {
+		// check for multiple refs returned
+		rbuf = bytes.NewReader(buf.Bytes())
+		err := json.NewDecoder(rbuf).Decode(&rs)
+		if err != nil {
+			return nil, resp, err
+		}
+	} else {
+		return nil, resp, err
+	}
+
+	return rs, resp, nil
```

Comment:
yes, that is much cleaner, and I like the preservation of the unmarshalling errors. code has been updated.